impl/src/main/java/org/jboss/weld/bean/DecoratedMethods.java

```diff
@@ -62,17 +67,37 @@
     }
 
     private Object findMatchingDecoratedMethod(Method method) {
+        // First try to find the same method
         for (InvokableAnnotatedMethod<?> decoratedMethod : decoratedTypeMethods) {
             if (decoratedMethod.getJavaMember().equals(method)) {
                 return decoratedMethod;
             }
         }
+        // Then try to find all matching methods
+        List<InvokableAnnotatedMethod<?>> matching = new ArrayList<InvokableAnnotatedMethod<?>>();
         for (InvokableAnnotatedMethod<?> decoratedMethod : decoratedTypeMethods) {
             if (matches(decoratedMethod, method)) {
-                return decoratedMethod;
+                matching.add(decoratedMethod);
             }
         }
-        return NULL_MARKER;
+        if (matching.isEmpty()) {
+            // No match
+            return NULL_MARKER;
+        } else if (matching.size() == 1) {
+            // If there is only one matching method no further action is needed
+            return matching.get(0);
+        }
+        // Choose the most specific method
+        // This does meet all requirements of JLS but it should work in most cases
```

Comment:
I do not fully understand the comment. Perhaps it is wrong? If this code meets all the JLS requirements then it should work in all (not most) cases, right?